runner.rb: make sure entire output is written to socket

sarahhodne opened the pull request:
The previous implementation would read and write a 1-byte buffer from the build.sh script to the socket, but the thread doing this is immediately closed once the process is finished. This makes it possible for some output not to have been read and written to the socket before the thread is closed, and we could lose some output at the end of a command.

The new implementation uses IO.copy_stream instead, which will copy the output from one IO object into another, and will block until the source object reaches an EOF. The current way of spawning the process with PTY.open and Process.spawn causes the "output" part of the pipe (the `io` variable) to not be closed at the end of the command, so that is changed to PTY.spawn, which does close the output IO object.

Commit:
runner.rb: make sure entire output is written to socket

The previous implementation would read and write a 1-byte buffer from
the build.sh script to the socket, but the thread doing this is
immediately closed once the process is finished. This makes it possible
for some output not to have been read and written to the socket before
the thread is closed, and we could lose some output at the end of a
command.

The new implementation uses IO.copy_stream instead, which will copy the
output from one IO object into another, and will block until the source
object reaches an EOF. The current way of spawning the process with
PTY.open and Process.spawn causes the "output" part of the pipe (the
`io` variable) to not be closed at the end of the command, so that is
changed to PTY.spawn, which does close the output IO object.

meatballhat reviewed (approved):
:+1: assuming there will be some testing done around the `stdin.close` bit

henrikhodne commented on bootstrap.sh:
```diff
@@ -183,17 +183,11 @@ require "socket"
-    end
-  end
+PTY.spawn("/usr/bin/env", "TERM=xterm", "/bin/bash", "--login", "/Users/travis/build.sh") do |stdout, stdin, pid|
+  stdin.close
```
I just tried this script on my Mac, and even with this line the script still stalled when trying to read something from stdin, so it appears it doesn't have any effect at all.

Commit:
runner.rb: remove stdin.close

After some testing it appears this doesn't actually do anything, so it's
better to remove it to avoid confusion.

henrikhodne commented:
Removed `stdin.close` since it didn't appear to do anything at all in my local testing.

emdantrim commented:
HECK YEAH
